### PR TITLE
Add xml parser

### PR DESCRIPTION
**Pull request by csh519**
For now CredSweeper scan xml file line by line.

So below case is detected now.
``` xml
<tool RefKey="DVmaxTimeTool" Name="EDT" />
```

It is FP case and CredSweeper should focus on value not attribute.
I want to make CredSweepr to detect below case.
``` xml
<?xml version='1.0' encoding='utf-8'?>
<Countries>
    <Country name="Korea">
        <City attr1="value1" attr2="value2">Seoul</City>
        <password name="password">cackle!</password>
    </Country>
    <Country name="Ukraine">
        <City attr1="value1" attr2="value2">Kyiv</City>
        <password name="password">peace_for_ukraine</password>
    </Country>
</Countries>
```

And here is change
- Add xml parser
- Add and modify tests

~*Actually there can be wrong line number on its result, because it is using xml parser so it is different with actual xml file.*~

### CODE REVIEW

**Comment by babenek** [comment]
```
Detected Credentials: 4490
Time Elapsed: 312.76039814949036s
result_cnt : 4086, lost_cnt : 67, true_cnt : 3688, false_cnt : 331
credsweeper -> TP : 3688, FP : 331, TN : 19429551, FN : 888, FPR : 0.0000170356, FNR : 0.1940559441, ACC : 0.9999372764, PRC : 0.9176412043, RCL : 0.8059440559, F1 : 0.8581733566
credsweeper Private Key -> TP : 952, FP : 0, TN : 4, FN : 39, FPR : None, FNR : 0.0393541877, ACC : 0.9608040201, PRC : 1.0000000000, RCL : 0.9606458123, F1 : 0.9799279465
credsweeper Predefined Pattern -> TP : 307, FP : 2, TN : 40, FN : 19, FPR : 0.0476190476, FNR : 0.0582822086, ACC : 0.9429347826, PRC : 0.9935275081, RCL : 0.9417177914, F1 : 0.9669291339
credsweeper Password -> TP : 980, FP : 129, TN : 4151, FN : 413, FPR : 0.0301401869, FNR : 0.2964824121, ACC : 0.9044597215, PRC : 0.8836789901, RCL : 0.7035175879, F1 : 0.7833733014
credsweeper Generic Secret -> TP : 978, FP : 4, TN : 214, FN : 77, FPR : 0.0183486239, FNR : 0.0729857820, ACC : 0.9363707777, PRC : 0.9959266802, RCL : 0.9270142180, F1 : 0.9602356406
credsweeper Generic Token -> TP : 278, FP : 5, TN : 598, FN : 55, FPR : 0.0082918740, FNR : 0.1651651652, ACC : 0.9358974359, PRC : 0.9823321555, RCL : 0.8348348348, F1 : 0.9025974026
credsweeper Other -> TP : 112, FP : 3, TN : 741, FN : 260, FPR : 0.0040322581, FNR : 0.6989247312, ACC : 0.7643369176, PRC : 0.9739130435, RCL : 0.3010752688, F1 : 0.4599589322
credsweeper Authentication Key & Token -> TP : 48, FP : 1, TN : 31, FN : 19, FPR : 0.0312500000, FNR : 0.2835820896, ACC : 0.7979797980, PRC : 0.9795918367, RCL : 0.7164179104, F1 : 0.8275862069
credsweeper Seed, Salt, Nonce -> TP : 33, FP : 0, TN : 8, FN : 6, FPR : None, FNR : 0.1538461538, ACC : 0.8723404255, PRC : 1.0000000000, RCL : 0.8461538462, F1 : 0.9166666667
```

**Comment by csh519** [comment]
@babenek 

Thank you for adding benchmark result.

File: credsweeper/file_handler/text_content_provider.py [hunk 1]
```diff
@@ -23,5 +24,13 @@ def get_analysis_target(self) -> List[AnalysisTarget]:
             list of analysis targets based on every row in file
 
         """
-        lines = Util.read_file(self.file_path)
+        if Util.get_extension(self.file_path) == ".xml":
+            try:
+                tree = ElementTree.parse(self.file_path)
+                lines = Util.get_xml_data(tree.getroot())
+            except ElementTree.ParseError:
```
Comment (babenek) [inline credsweeper/file_handler/text_content_provider.py:31]: Possibly common Exception should be caught? 
There may be encoding error. Not sure it wrapped in xml parser

File: credsweeper/file_handler/text_content_provider.py [hunk 2]
```diff
@@ -23,5 +24,13 @@ def get_analysis_target(self) -> List[AnalysisTarget]:
...
+                lines = Util.read_file(self.file_path)
+        else:
+            lines = Util.read_file(self.file_path)
```
Comment (babenek) [inline credsweeper/file_handler/text_content_provider.py:34]: Duplicates...
Probably need to add an error_flag and split...

Comment (babenek) [inline credsweeper/file_handler/text_content_provider.py:34]:
```suggestion
        error = True
        if Util.get_extension(self.file_path) == ".xml":
            try:
                tree = ElementTree.parse(self.file_path)
                lines = Util.get_xml_data(tree.getroot())
                error = False
            except Exception as exc:
                logger.error(f"Parse xml fail:{exc}")
        if error:
            lines = Util.read_file(self.file_path)
```

File: tests/utils/test_util.py [hunk 3]
```diff
@@ -274,3 +275,13 @@ def test_is_gzip_n(self):
...
         self.assertFalse(Util.is_gzip(b'\x1f\x8bxxx'))
         self.assertFalse(Util.is_gzip(b'\x1f\x8b\x02'))
+
+    def test_get_xml_data_p(self):
+        target_path = str(SAMPLES_DIR / "xml_password.xml")
+        tree = ElementTree.parse(target_path)
+        lines = Util.get_xml_data(tree.getroot())
+
+        assert lines == [
+            "Countries : \n    ", "Country : \n        ", "City : Seoul", "password : cackle!", "Country : \n        ",
```
Comment (babenek) [inline tests/utils/test_util.py:285]: \n - are not stripped?

**Review by babenek — COMMENTED** [review]
Please, review suggestion. May be other file types should be added to parse.

File: tests/utils/test_util.py (hunk 3 above)
Comment (csh519) [reply tests/utils/test_util.py]: Good point.
I fixed it with 263291b.

File: credsweeper/file_handler/text_content_provider.py (hunk 2 above)
Comment (csh519) [reply credsweeper/file_handler/text_content_provider.py:34]: Thank you for suggestion!
Sadly `if error:` is not proper here... i think..
I'll refactor this code block.

Comment (csh519) [reply credsweeper/file_handler/text_content_provider.py:42]: Added `lines_init` flag [dbd77cd](https://github.com/Samsung/CredSweeper/pull/193/commits/dbd77cd6d36e6cf3457ed27018b881535471c40a).
Thank you for review.

**Comment by babenek** [comment]
please, check the case:
```
<AnElement
            android:defaultValue="false"
            app:key="password"
            app:somethingElse="dfhgkjdgh"/>
```
Possibly it throws ``Cannot parse '..... .xml' to xml 'NoneType' object has no attribute 'strip'
and this (maybe not an issue):
```
<item name="colorPrimary">@color/colorPrimary</item>
```
Cannot parse 't.xml' to xml 'cython_function_or_method' object has no attribute 'strip'

**Comment by csh519** [comment]
@babenek 

Regarding first case, it is right operation.
There is no prefix for `android` and `app`.
So it is impossible to parse the example.

-----

And the second case, there is no error log when I test this case.
Did you check if the right version of `lxml` was installed? (requirements.txt)

**Comment by babenek** [comment]
@csh519 credsweeper was installed with ``pip install .`` and version of lxml is 4.9.1

**Comment by csh519** [comment]
@babenek 

It's weird.. I can't reproduce the error you told..
Could you test it again with latest pull of this branch?

**Comment by babenek** [comment]
Well, possibly not the case. Check the files:
https://github.com/google/google-authenticator/blob/51781910ae2bb1abf8ac51b290272f86f3651235/mobile/blackberry/build.xml
https://github.com/iceleo-com/vinakey/blob/9499f43626fcbdcd45e7474adae268d7c4cbc981/app/src/main/res/values/themes.xml

**Comment by csh519** [comment]
@babenek 
I fixed it by c565536. Now all cases that you told are executed successfully.
Please check.

File: credsweeper/file_handler/text_content_provider.py [hunk 4]
```diff
@@ -23,5 +26,13 @@ def get_analysis_target(self) -> List[AnalysisTarget]:
...
         """
-        lines = Util.read_file(self.file_path)
+        lines = None
+
+        if Util.get_extension(self.file_path) == ".xml":
+            lines, line_nums = Util.get_xml_data(self.file_path)
+            if lines:
+                return self.lines_line_num_to_targets(lines, line_nums)
+
+        if lines is None:
```
Comment (babenek) [inline credsweeper/file_handler/text_content_provider.py:36]: 'if' may be extra due there was return before

**Comment by babenek** [comment]
Ok. The issues vanished. Benchmark updated 
https://github.com/babenek/CredSweeper/commit/fe178c99b3c61885f153709785a7acced65eeb07

File: credsweeper/file_handler/text_content_provider.py (hunk 4 above)
Comment (csh519) [reply credsweeper/file_handler/text_content_provider.py:35]: Actually it is needed.
But I refactored the code for more readability.
Please check f02995f.

File: credsweeper/utils/util.py [hunk 5]
```diff
@@ -270,6 +272,53 @@ def read_data(path: str) -> Optional[bytes]:
...
+        Return:
+            List of formatted string(f"{root.tag} : {root.text}")
+
+        """
+        lines = []
+        line_nums = []
+        try:
+            with open(file_path, "r") as f:
+                xml_lines = f.readlines()
+            tree = etree.fromstringlist(xml_lines)
```
Comment (babenek) [inline credsweeper/utils/util.py:293]:
```suggestion
            tree = etree.parse(file_path)
```
May we use etree's file reading? I've not found xml in UTF-16 but it seems possible.

File: credsweeper/utils/util.py [hunk 6]
```diff
@@ -270,6 +272,53 @@ def read_data(path: str) -> Optional[bytes]:
...
+            element: xml element
+            attr: attribute name
+
+        Return:
+            String xml data with strip()
+
+        """
+        if getattr(element, attr) is None or not isinstance(getattr(element, attr), str):
+            return ""
+        return str(getattr(element, attr)).strip()
```
Comment (babenek) [inline credsweeper/utils/util.py:320]:
```suggestion
        element_attr = getattr(element, attr)
        if element_attr is None or not isinstance(element_attr, str):
            return ""
        return element_attr.strip()
```
a bit optimization(?) not sure whether python does it

File: credsweeper/utils/util.py (hunk 5 above)
Comment (csh519) [reply credsweeper/utils/util.py:293]: Actually I used `etree.parse()`, but there is no `sourceline` attribute. It is needed to recognize exact line number of xml file.
So I change it to use `etree.fromstring()`.

File: credsweeper/utils/util.py (hunk 6 above)
Comment (csh519) [reply credsweeper/utils/util.py]: Thank you for suggestion. I refactored the code with f00c5e1.
Please check.

**Review by kmnls — APPROVED** [review]

**Review by babenek — APPROVED** [review]
LGTM.
Benchmark scores were changed

**Pull request merged**